experimental/oak_async/src/executor.rs

```diff
@@ -0,0 +1,176 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//! Async executor for Oak Nodes.
+
+use core::{cell::RefCell, future::Future, task::Waker};
+use futures::{executor::LocalPool, task::LocalSpawnExt};
+use log::{debug, trace};
+use oak::{ChannelReadStatus, Handle, OakStatus, ReadHandle};
+use std::collections::HashMap;
+
+// thread local so `cargo test` can run tests in parallel.
+// Oak nodes are single-threaded.
```

Comment:
Is this going to break when Wasm supports multi threaded nodes? It seems a bit of a hack if it's just for testing.

Reply:
When wasm gains support for multithreading, I think we'll still want to have this thread local variable, but the inner value will probably be backed by an `Arc` so the actual object is global. When a new thread is created it will inherit the executor from its parent. That way tests will still work, and you could even run multiple executors in different threads, each with their own threadpool.

It's worth noting Tokio also uses a thread local variable to store the handle to an executor: https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/context.rs#L7.